src/ruby/puppet-server-lib/puppet/server/master.rb

```diff
@@ -57,6 +60,53 @@ def handleRequest(request)
         response["X-Puppet-Version"])
   end
 
+  def getClassInfoForAllEnvironments()
+    # The clear_environment_settings(env) ensures that the calls which happen
+    # later on to enumerate all of the manifests for each environment are
+    # using the latest environment.conf settings for each environment.  Without
+    # this call, cached (and, therefore, possibly stale) environment.conf
+    # settings would be used.  We want to return the latest data for any calls
+    # made to this method.
+    @env_loader.list.each do |env|
+      Puppet.settings.clear_environment_settings(env)
+    end
+
+    environments =
+      Hash[@env_loader.list.collect do |env|
+       [env.name, self.class.getManifests(env)]
+      end]
```

Comment:
@cprice404 It's not just to get the "names" of the environments.  It also iterates through the available {{Puppet::Node::Environment}} objects to get all of the manifest files that need to be parsed for each environment, building the result into a Hash that is passed into the new {{Puppet::InfoService::ClassInformationService}} to parse out the class info from all of those manifests.  And a key piece of that is ensuring that the list which is returned contains "latest data" - i.e., would include all of the environments currently on disk and also use the {{environment.conf}} settings as currently specified on disk.

By "new API" what are you thinking of as being the scope?  Would that API essentially just allow the code in Puppet Server here to make some sort of {{getManifests}} call down into core Ruby Puppet code, returning the Hash of environment => manifest files from "latest data" -- essentially having all of the work to enumerate manifests from the environments be done behind that new {{getManifests}} API?  Or were you just thinking of some sort of construct that would allow for a list of {{Puppet::Node::Environment}} objects compromising "latest data" that can be mined out like is done here behind the `self.class.getManifests` call?

Or - going even further - were you thinking of just thinning this logic out almost entirely to where all this does is a passthrough to core Ruby Puppet code that does the manifest enumeration and parsing via the `ClassInformationService` all in one step?